### PR TITLE
spec: support composite query

### PR DESCRIPTION
**Pull request by chenyan-dfinity**
Follow up with https://github.com/dfinity/interface-spec/pull/144

### CODE REVIEW

File: spec/Candid.md
Comment (ggreif) [inline spec/Candid.md:73]: I still think this is a very badly chosen name. My suggestion (that came too late to be considered in the replica naming contest) was _report_.

Comment (chenyan-dfinity) [reply spec/Candid.md:73]: I wish it's named `query_tmp_hack`

Comment (nomeata) [reply spec/Candid.md:73]: QUERY. As in, a larger query.

Comment (chenyan-dfinity) [reply spec/Candid.md:73]: QUEyr. It's bigger than `query`, but not big enough (only the same subnet). The spelling is wrong, because it has strange await semantics...

File: spec/Candid.md
Comment (ulan) [inline spec/Candid.md:194]: > cannot be called in inter-canister calls
I think it should be "cannot be called by update methods" because it is possible to make an inter-canister composite query call.

Could you please also add that the restriction about the "update method" and "same subnet" are due to the incompleteness of current implementation and may be lifted in the future?

File: spec/Candid.md
Comment (ulan) [reply spec/Candid.md:73]: @ggreif: I associate "report" with documents and graphs, not queries. If I was a developer, I would be confused by that name. I think the name has to contain the word "query" or at the very least "queyr" 😄 

@chenyan-dfinity: "because it has strange await semantics..."
There is a trade-off between strange await semantics and strange query semantics. Community voted to keep the proper query semantics sacrificing a corner case of self-call await. We also surveyed the use cases of people interested in composite queries and all of the uses cases involved calling different canisters, where await works as expected.

File: spec/Candid.md
Comment (ggreif) [reply spec/Candid.md:194]: Here are my further naming suggestions
- `inquisition` (religious undertone)
- `interrogation` (too long, menacing)
- `quiz` (many questions, sounds fitting)

File: spec/Candid.md
Comment (ggreif) [reply spec/Candid.md:73]: Here are my further naming suggestions
- `inquisition` (religious undertone)
- `interrogation` (too long, menacing)
- `quiz` (many questions, sounds fitting)

Comment (nomeata) [reply spec/Candid.md:73]: We could just ask Dom. That's how naming debates have been resolved in the past :-)

**Review by crusso — APPROVED** [review]
But do address @Ulan's suggestions.

File: spec/Candid.md
Comment (crusso) [reply spec/Candid.md:73]: I still quite like `enquiry` or `inquiry`, but never mind.

Comment (ulan) [reply spec/Candid.md:73]: I hope this discussion is not blocking the PR.

File: spec/Candid.md
Comment (ulan) [reply spec/Candid.md:194]: I am approving assuming this comment will be addressed.

**Review by ulan — APPROVED** [review]
Lgtm, thanks!

File: spec/Candid.md
Comment (chenyan-dfinity) [reply spec/Candid.md:194]: > may be lifted in the future

Lifting all restrictions means we will have just the `query` method, and no `composite query` method, or it's the other way around?

Candid spec is quite high level, and tries to be platform agnostic. See the paragraph above about `query` method. It's only suggesting that `query` method can be more efficient. I prefer to describe `composite query` without mentioning subnet, but seems like it's not easy?

Comment (ulan) [reply spec/Candid.md:194]: Sorry for the delay. I thought I've replied already, but don't see my reply here.

> Lifting all restrictions means we will have just the query method, and no composite query method, or it's the other way around?

Query and composite query will become the same thing:

- query will be able to call query and composite query
- composite query will be callable from an update

> Candid spec is quite high level, and tries to be platform agnostic. See the paragraph above about query method. It's only suggesting that query method can be more efficient.

It seems inconsistent that we mention limitations of composite queries without mentioning the main limitation of query: it cannot call any other methods.

If we want to stay platform agnostic, then maybe we should write something more vague: "composite query is a kind of a query with additional platform-specific features and limitations"?

That's said, my personal preference would be to be more pragmatic and mention all information that might be useful for the users (maybe with some comment that it only applies to IC): 
- composite queries can be called only from other composite queries.
- composite queries can call composite queries and queries.
- composite query calls cannot be made cross-subnets.
- queries cannot call any other methods.
All these limitations are temporary due to the implementation.

Comment (chenyan-dfinity) [reply spec/Candid.md]: I made another pass. PTAL.

> composite queries can be called only from other composite queries.

It can also be called outside of IC.

> queries cannot call any other methods.

There used to be a hack that a query function can call another terminating query function. Is this being removed?

Comment (ulan) [reply spec/Candid.md]: Thanks, the latest version lgtm!

> There used to be a hack that a query function can call another terminating query function. Is this being removed?

Do you mean ICQC on system and verified application subnets? There non-replicated queries can call other queries on the same subnet. That is exactly what we are packaging as composite_query and releasing on all subnets.

Right now, queries on application subnets cannot call any other queries (terminating or not).

**Pull request merged**
